src/domain/users/user-organizations.repository.ts

```diff
@@ -0,0 +1,288 @@
+  implements IUsersOrganizationsRepository
+{
+  constructor(
+    private readonly postgresDatabaseService: PostgresDatabaseService,
+    @Inject(IUsersRepository)
+    private readonly usersRepository: IUsersRepository,
+    @Inject(IOrganizationsRepository)
+    private readonly organizationsRepository: IOrganizationsRepository,
+    @Inject(IWalletsRepository)
+    private readonly walletsRepository: IWalletsRepository,
+  ) {}
+
+  public async findOneOrFail(
+    where:
+      | Array<FindOptionsWhere<UserOrganization>>
+      | FindOptionsWhere<UserOrganization>,
+    relations?: FindOptionsRelations<UserOrganization>,
+  ): Promise<DbUserOrganization> {
+    const organization = await this.findOne(where, relations);
+
+    if (!organization) {
+      throw new NotFoundException('User organization not found.');
+    }
+
+    return organization;
+  }
+
+  public async findOne(
+    where:
+      | Array<FindOptionsWhere<UserOrganization>>
+      | FindOptionsWhere<UserOrganization>,
+    relations?: FindOptionsRelations<UserOrganization>,
+  ): Promise<DbUserOrganization | null> {
+    const organizationRepository =
+      await this.postgresDatabaseService.getRepository(DbUserOrganization);
+
+    return await organizationRepository.findOne({
+      where,
+      relations,
+    });
+  }
+
+  public async inviteUsers(args: {
+    authPayload: AuthPayload;
+    orgId: Organization['id'];
+    users: Array<{
+      address: `0x${string}`;
+      role: UserOrganizationRole;
+    }>;
+  }): Promise<
+    Array<{
+      userId: User['id'];
+      orgId: Organization['id'];
+      role: keyof typeof UserOrganizationRole;
+      status: keyof typeof UserOrganizationStatus;
+    }>
+  > {
+    const signer = await this.getOrgAndSignerOrFail(args);
+
+    this.assertUserOrgIsActive(signer.userOrg);
+
+    const invitations: Array<{
+      userId: User['id'];
+      orgId: Organization['id'];
+      role: keyof typeof UserOrganizationRole;
+      status: keyof typeof UserOrganizationStatus;
+    }> = [];
+
+    await this.postgresDatabaseService.transaction(async (entityManager) => {
+      for (const userToInvite of args.users) {
+        const invitedUser = await this.walletsRepository.findOneByAddress(
+          userToInvite.address,
+          { user: true },
+        );
+
+        let invitedUserId = invitedUser?.id;
+
+        if (!invitedUserId) {
+          invitedUserId = await this.usersRepository.create(
+            UserStatus.PENDING,
+            entityManager,
+          );
+
+          await this.walletsRepository.create(
+            {
+              walletAddress: userToInvite.address,
+              userId: invitedUserId,
+            },
+            entityManager,
+          );
+        }
+
+        await entityManager.insert(DbUserOrganization, {
+          user: { id: invitedUserId },
+          organization: signer.org,
+          role: userToInvite.role,
+          status: UserOrganizationStatus.INVITED,
+        });
```

Comment:
What happens when an admin invites two wallets that belong to the same user. Would this request throw for the second wallet since we are trying to insert the same user id for the same org into the user-org table?

Reply:
We discussed this and initially we thought as long as we do an upsert it should be fine since for the second wallet we would just update the existing membership entry of the user but now that I think about it again what do we do if those wallets were invited with different roles? It might be highly unlikely but still leaves room for interpretation (do we keep the initial role or always also update the role again -> behaviour is not clear to the admin who invites wallets).

Overall I think if we limited this flow to single wallet invites it would make error handling a lot easier since we could detect if that user already got invited and show an error message or update the role if its different. I will bring this up in the next sync.